tsl/src/continuous_aggs/refresh.c

```diff
@@ -342,7 +357,37 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 	 */
 	LockRelationOid(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
 					AccessExclusiveLock);
-	continuous_agg_invalidation_threshold_set(cagg->data.raw_hypertable_id, refresh_window.end);
+
+	/* Compute new invalidation threshold. Note that this computation caps the
+	 * threshold at the end of the last bucket that holds data in the
+	 * underlying hypertable. */
+	computed_invalidation_threshold = invalidation_threshold_compute(cagg, &refresh_window);
+
+	/* Set the new invalidation threshold. Note that this only updates the
+	 * threshold if the new value is greater than the old one. Otherwise, the
+	 * existing threshold is returned. */
+	invalidation_threshold = invalidation_threshold_set_or_get(cagg->data.raw_hypertable_id,
+															   computed_invalidation_threshold);
+
+	/* We must also cap the refresh window at the invalidation threshold. If
+	 * we process invalidations after the threshold, the continuous aggregates
+	 * won't be refreshed when the threshold is moved forward in the
+	 * future. The invalidation threshold should already be aligned on bucket
+	 * boundary. */
+	if (invalidation_threshold <= refresh_window.end)
```

Comment:
```suggestion
	if (refresh_window.end => invalidation_threshold)
```
then it is the same order as the next line and thus easier to follow.